Auto Updating Pipeline for web app

danich1 opened the pull request:
This PR contains code for automatically incorporating new PubMed Central OA (PMCOA) papers into our web service. Note: the code style here has been formatted by black, so not much wiggle room for stylistic changes. Also feel free to skip scripts that aren't in the auto_updater folder as I just ran black with no substantial changes.

Anyway the pipeline works is as follows:
1. Opens a network stream to PMCOA's ftp server
2. Scans the tar.gz files for any paper that's "new" to our dataset
3. Extracts/Parses the paper in real time and writes new information to files
4. Once extraction/parsing is done it updates the existing files the backend server uses along with the front end as well

The file paths here are based on my local machine; however, I expect them to change based on the google cloud instance you have set up. Wanted to get your feedback on that once you have a chance to look at my code. Also, adding a note that we need to update some files on the google cloud instance as they are quite smaller than originally thought.

dongbohu commented on server/auto_updater/auto_updater_runner.py:
```diff
@@ -0,0 +1,87 @@
+    update_dictionaries,
+)
+
+from embedding_bin_updater import (
```
Did you forget to add `embedding_bin_updater.py` to this PR? I can't find this module.

dongbohu commented on server/auto_updater/auto_updater_runner.py:
```diff
@@ -0,0 +1,87 @@
+    temp_token_count_filename = "local_data/new_paper_token_count.tsv"
+
+    # Current paper files
+    current_pmc_dir_filename = "local_data/pmc_oa_file_list.tsv"
```
Can't find this file in the repo. Can you add it either to the repo or to the Google cloud bucket?

danich1 replied:
Since some of the files are more gigs than github allows, I'll place these files onto the google cloud instance.

dongbohu commented on server/auto_updater/auto_updater_runner.py:
```diff
@@ -0,0 +1,87 @@
+    # Current paper files
+    current_pmc_dir_filename = "local_data/pmc_oa_file_list.tsv"
+    paper_dataset_filename = "local_data/paper_dataset_full.tsv.xz"
+    paper_landscape_file = "local_data/paper_dataset_tsne_square.tsv"
```
Can't find this file in the repo either. Can you add it either to the repo or to the Google cloud bucket?

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,273 @@
+    with open(temp_dir_filename, "w") as dir_file, open(
+        temp_embed_filename, "w"
+    ) as embed_file:
+        with open(temp_token_counter, "w") as word_file:
```
This `with` clause can probably merged with the one on lines 68-70. Up to you.

dongbohu commented on server/auto_updater/get_square_bins.R:
```diff
@@ -0,0 +1,9 @@
+library(ggplot2)
```
Can you tell me what this script is for? I didn't find it invoked anywhere.

danich1 replied:
This script is used to generate the square bin plot that you see on the main page of the web app. There aren't any python packages that can generate this plot while allowing the user to access each individual square bin. Hence the need to use R in this situation.

dongbohu reviewed (commented):
Left a few comments. Need some input files to test the memory usage.

danich1 commented:
The local files are now uploaded onto google cloud with the exception being: pmc_square_plot.json. This is part of the repository itself so I don't think I needed to duplicate the file.

dongbohu commented:
@danich1: I got the following error when processing `comm_use.A-B.xml.tar.gz`:
```
Requesting ftp://ftp.ncbi.nlm.nih.gov/pub/pmc/oa_bulk/comm_use.A-B.xml.tar.gz....
  6%|██████▎                                                                                              | 1/16 [00:05<01:22,  5.48s/it]
Traceback (most recent call last):
  File "auto_updater_runner.py", line 42, in <module>
    gather_new_papers(
  File "/home/ubuntu/pss/server/auto_updater/dataset_updater.py", line 137, in gather_new_papers
    for tok in word_counter:
TypeError: 'NoneType' object is not iterable
```
Maybe we should add the following lines before this line: 
https://github.com/danich1/preprint-similarity-search/blob/c83b7fab20d72afe35da39c4c6f773f2d7888ab0/server/auto_updater/dataset_updater.py#L137
```
if word_counter is None:
    break
```

dongbohu commented on server/requirements.txt:
```diff
@@ -1,15 +1,21 @@
+black==20.8b1
```
This is for development only. We probably shouldn't include it.

dongbohu commented on server/requirements.txt:
```diff
@@ -1,15 +1,21 @@
 flask==1.1.2
 flask-cors==3.0.8
 flask_restful==0.3.8
 gensim==3.8.2
```
can you also upgrade `gensim` to `3.8.3` (which is a bugfix release)? See https://github.com/RaRe-Technologies/gensim/releases/tag/3.8.3

dongbohu commented on server/requirements.txt:
```diff
@@ -1,15 +1,21 @@
 flask-cors==3.0.8
 flask_restful==0.3.8
 gensim==3.8.2
+gsutil==4.59
```
This is for deployment only. Not sure whether it should be included.

danich1 replied:
Well I was under the impression this package would be necessary for accessing data on the google cloud instance. Is the plan to run this code on the instance itself?

dongbohu replied:
Where to run the cron update depends on the memory usage of the code. Let's keep this package for now.

dongbohu commented on server/requirements.txt:
```diff
@@ -1,15 +1,21 @@
 spacy==3.0.3
 tables==3.6.1
 tensorflow==2.4.1
+tqdm==4.56.2
```
let's also include `wheel`, which is needed for installations of a few packages. (I just realized that `python -m venv` doesn't include it by default but `virtualenv` command does.)

dongbohu commented:
@danich1 I got the following error now:
```
Requesting ftp://ftp.ncbi.nlm.nih.gov/pub/pmc/oa_bulk/comm_use.A-B.xml.tar.gz....
  6%|███████▏                                                                                                           | 1/16 [03:10<47:40, 190.72s/it]
Traceback (most recent call last):
  File "auto_updater_runner.py", line 42, in <module>
    gather_new_papers(
  File "/home/ubuntu/pss/server/auto_updater/dataset_updater.py", line 107, in gather_new_papers
    pmc_paper = open_stream.next()
  File "/usr/lib/python3.8/tarfile.py", line 2307, in next
    raise ReadError("unexpected end of data")
tarfile.ReadError: unexpected end of data
```
Any idea why?

danich1 commented:
>Any idea why? 

Could be that the stream was corrupted while you were running my code or there is an error in a compressed file I was trying to parse. If you get that issue again, I'll wrap that statement in a try catch exception.

dongbohu commented:
@danich1 Got that error again. Can you add the `try/except` block? Thanks.

dongbohu commented:
@danich1: I tested the script on a virtual machine whose RAM is 8 GB (same size as the API production server). It crashed when updating centroid dataset. Obviously the `generate_centroid_dataset()` in `embedding_bin_updater.py` consumed a lot of memory because it loads the whole paper dataset into pandas. Is there any way to decrease the memory usage of this function (even if it would run slower)?

dongbohu commented:
By the way, what's the RAM size of your local machine? How long did the script take on your local machine?

danich1 commented:
>By the way, what's the RAM size of your local machine? How long did the script take on your local machine?

64GB hence why I'm not surprised it broke a 8GB machine. Unfortunately, I only stress tested for one document on there, so I'll need to run it myself to answer the last question.

danich1 commented:
>By the way, what's the RAM size of your local machine? How long did the script take on your local machine?

Yes, I have a way to make it more memory efficient; however, I'll need time to code/test it. Will update once I finish that task

dongbohu commented:
@danich1 Got this error now:
```
Traceback (most recent call last):
  File "auto_updater_runner.py", line 68, in <module>
    generate_centroid_dataset(paper_dataset_filename, centroid_dataset_filename)
  File "/home/ubuntu/pss/server/auto_updater/embedding_bin_updater.py", line 52, in generate_centroid_dataset
    [float(line[f"feat_{idx}"]) for idx in range(dim)]
  File "/home/ubuntu/pss/server/auto_updater/embedding_bin_updater.py", line 52, in <listcomp>
    [float(line[f"feat_{idx}"]) for idx in range(dim)]
ValueError: could not convert string to float: ''
```

danich1 commented:
what are the values of the line variable? This error says that one of the feature entry is null, which shouldn't be possible if papers are being embedded correctly.

dongbohu commented:
@danich1 I have to rerun it from the beginning to reproduce the error. Since you will run the whole process locally, let's see whether you can reproduce the error too.

dongbohu commented:
@danich1 I tested the latest code that you committed last night. The centroid was updated successfully, but "Updating Token/Token bin Counts" stage crashed due to the size of RAM on the my test box. How was your local run? Is it done w/o any error?

danich1 commented:
>@danich1 I tested the latest code that you committed last night. The centroid was updated successfully, but "Updating Token/Token bin Counts" stage crashed due to the size of RAM on the my test box. How was your local run? Is it done w/o any error?

Yeah I'm not surprised. Token counts take up quite a bit of memory. My code ran on my desktop as it has 64GB of RAM. 🤷🏾‍♂️ on any fix for a 8 GB machine. If you have any ideas, please let me know. Also, warning the last function takes up to 12 hours so that is a thing as well.

dongbohu commented:
@danich1 Thanks for the update. How long did the updater take on your 64-GB-RAM local machine totally?

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,285 @@
+
+        object_to_be_updated = pickle.load(
+            open(
+                f"bin_counters/word_bin_{bin_num_str + str(square_bin)}_count.pkl", "rb"
```
Where is this `bin_counters/` directory (and the pickle files)? I can't find it in either the repo and Google cloud bucket.

danich1 replied:
Forgot to add it. Will upload shortly.

danich1 replied:
Uploaded!!

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,285 @@
+            )
+        )
+        object_to_be_updated.update(token_bin_dictionaries[square_bin])
+        pickle.dump(
```
This line seems to dump the object back to the same pickle file that was loaded on line 276. Should we close the file handle before this line?

danich1 commented:
>@danich1 Thanks for the update. How long did the updater take on your 64-GB-RAM local machine totally?

Need to confirm, but roughly 13-14 hours at most. The final function is the slowest.

dongbohu commented on server/auto_updater/embedding_bin_updater.py:
```diff
@@ -0,0 +1,319 @@
+    (
+        square_bin_plot_df.assign(bin_id=mapped_data_df.squarebin_id.unique().shape[0])[
+            ["x", "y", "xmin", "xmax", "ymin", "ymax", "bin_id"]
+        ].to_json("pmc_square_plot.json", orient="records", lines=False)
```
I think `"pmc_square_plot.json"` should be replaced by `paper_landscape_json_file`.

dongbohu commented:
@danich1: One question on the files that you uploaded to the Google cloud bucket: are they supposed to be input files or output files of the automatic updater? I found it a little confusing because these files were overwritten by the updating process.

danich1 commented:
>@danich1: One question on the files that you uploaded to the Google cloud bucket: are they supposed to be input files or output files of the automatic updater? I found it a little confusing because these files were overwritten by the updating process.

Re-posting for the public. The input files are designed to be updated as part of the process. Didn't want to deal with duplicating large files, which resulted in the idea of appending input files with new data. Good news is the input files on google cloud are the original files not the appended ones.

dongbohu commented:
Thanks @danich1. The good thing of modifying the files in place is that it saves disk space. The bad thing is that when something goes wrong during updating, we have to undo the changes before restarting. I am going to run another test today.

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,288 @@
+        for tar_file in tqdm.tqdm(tar_files):
+
+            # If not xml files skip
+            if all([suffix != ".xml" for suffix in Path(tar_file).suffixes]):
```
The `[]` inside `()` is unnecessary, so this line can be simplified a little bit , like this:
```if all(suffix != ".xml" for suffix in Path(tar_file).suffixes):```

I noticed that sometimes the FTP site includes a `*.tmp` file, for example, this is the list of files that I found this morning:
```
comm_use.0-9A-B.txt.tar.gz
comm_use.A-B.xml.tar.gz
comm_use.C-H.txt.tar.gz
comm_use.C-H.xml.tar.gz
comm_use.I-N.txt.tar.gz
comm_use.I-N.xml.tar.gz
comm_use.I-N.xml.tar.gz.tmp
comm_use.O-Z.txt.tar.gz
comm_use.O-Z.xml.tar.gz
non_comm_use.0-9A-B.txt.tar.gz
non_comm_use.A-B.xml.tar.gz
non_comm_use.C-H.txt.tar.gz
non_comm_use.C-H.xml.tar.gz
non_comm_use.I-N.txt.tar.gz
non_comm_use.I-N.xml.tar.gz
non_comm_use.O-Z.txt.tar.gz
non_comm_use.O-Z.xml.tar.gz
```
Note the extra file `comm_use.I-N.xml.tar.gz.tmp`? It existed in the FTP site for a while, then disappeared. If you check their website: here: https://ftp.ncbi.nlm.nih.gov/pub/pmc/oa_bulk/
You can see that `comm_use.I-N.xml.tar.gz` was updated this morning. So my guess is that this extra file was generated when they updated the website. When updating is done, the tmp file replaces the original one. If my guess is correct, we should ignore "*.tmp" files, because they are incomplete.

danich1 commented:
>@danich1 Are you commenting out the lines 39-79 for testing purpose?

Oops meant to remove that before pushing new code. Fix it now!

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,288 @@
+                        f"File {pmc_open_access_url}{tar_file} achieved an error: {str(e)}"
+                    )
+                    print("Skipping...")
+                    break
```
It seems to make more sense to use `continue` on this line (or delete this line completely), because the exception happens on the current `pmc_paper`, but we may still be able to process the other papers in this tarball file.

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,292 @@
+                            }
+                        )
+
+                except tarfile.ReadError as e:
```
@danich1: After doing a few more tests over the weekend, I don't think this exception should be silently swallowed. To give you an idea: this exception was thrown on my test box after processing **1,084** xml files in `comm_use.A-B.xml.tar.gz`, but this tarball file includes totally **385,048** xml files, so most of them were ignored when the `break` line on 156 went out of `while` loop. 

I also downloaded `comm_use.A-B.xml.tar.gz` onto my hard drive and run `generate_new_papers()` function on this tarball file only. It took more than 10 hours and found 63,192 new papers. You can modify the function and test it by yourself. Since the FTP site includes 8 xml tarball files and their total size is over 50GB, the step to generate new papers is going to be much longer than expected, unless we use multi-thread or multi-process to speed up. I need to do more tests on this step.

When you ran the whole process on your local computer, did you ever gotten this exception? How did did it take to finish the `generate_new_papers()` step? How many new papers did it eventually find?

danich1 replied:
Oh man! 

>When you ran the whole process on your local computer, did you ever gotten this exception? 

Yup I did. Hence why I added the exception with a break statement at the end. I though skipping and waiting for PMCOA peeps to fix this file would be sufficient. Assumed the file was just broken.

>How did did it take to finish the generate_new_papers() step?
> How many new papers did it eventually find?

Like 30-45 minutes since the amount of new papers were tiny (1221 new papers). Given this discovery I agree multi-processing is needed for a needed speed up on this function along with: `update_paper_bins_stats()`

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,292 @@
+import spacy
+import tqdm
+
+nlp = spacy.load("en_core_web_sm")
```
@danich1: I forgot to tell you: if you want to run `generate_new_papers()` function on the locally downloaded `comm_use.A-B.xml.tar.gz`, you need to add the following line right after line 15:
```nlp.max_length = 1600000```
otherwise you would get the following error after about 2.5 hours:
```
Traceback (most recent call last):
  File "./1_generate_new_papers.py", line 174, in <module>
    dhu_generate_new_papers(
  File "./1_generate_new_papers.py", line 90, in dhu_generate_new_papers
    doc_vector, word_counter = generate_vector_counts(
  File "./1_generate_new_papers.py", line 151, in generate_vector_counts
    nlp(all_text),
  File "/home/dhu/.venv/pss/lib/python3.8/site-packages/spacy/language.py", line 977, in __call__
    doc = self.make_doc(text)
  File "/home/dhu/.venv/pss/lib/python3.8/site-packages/spacy/language.py", line 1056, in make_doc
    raise ValueError(
ValueError: [E088] Text of length 1347496 exceeds maximum of 1000000. The parser and NER models
require roughly 1GB of temporary memory per 100,000 characters in the input. This means long
texts may cause memory allocation errors. If you're not using the parser or NER, it's probably
safe to increase the `nlp.max_length` limit. The limit is in number of characters, so you can
check whether your inputs are too long by checking `len(text)`.
```
I suspect that setting `nlp.max_length` to any number larger than 1347496 may work, but I used `1600000` to err on the side of caution.

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,292 @@
+            filter(
+                lambda tok: str(tok) in model.wv
+                and str(tok) not in nlp.Defaults.stop_words,
+                nlp(all_text),
```
@danich1: I measured the time of each step in this function carefully and found that this line is the bottleneck. It took about 0.6-0.8 seconds to run (depending on the size of the ppaer in XML format), which creates a `nlp` instance. After reading this page: https://spacy.io/usage/processing-pipelines#disabling
I changed line 15 from:
```
nlp = spacy.load("en_core_web_sm")
``` 
to:
```
disabled_pipelines = [
    "tagger",
    "parser",
    "ner",
    "lemmatizer",
    "attribute_ruler",
    "tok2vec",
]
nlp = spacy.load("en_core_web_sm", disable=disabled_pipelines)
```
which basically disabled all default pipelines. This change improved the processing time dramatically. Now the new papers in `comm_use.A-B.xml.tar.gz` was generated in 40 minutes (instead of more than 10 hours). I then compared the three output files: 
1. `new_papers_added_dir.tsv` is identical, this is not surprising. 
2. `new_papers_embedding.tsv` is slightly diffferent for one paper (out of 63,192 papers):
```
# "new_papers_embedding.tsv" by nlp that has no any pipelines:
BMC_Med_Genomics	PMC7011249	-0.13292737	0.29884017	-0.44909483	-0.10702868	-0.2728784	-0.27431023	0.04570517	-0.32943398	0.25982034	-0.30733883	-1.0842022	-0.31371504	0.08960449	0.048983153	-0.2921648	0.21691155	0.36423025	-0.18740693	-0.22824855	0.18023007	0.1829937	-0.17935291	0.43323973	0.5302209	0.11441889	-0.20402277	-0.71329224	-0.3292653	-0.42242202	-0.124855444	0.38212383	-0.35289082	-0.43809974	-0.040971376	0.03614347	-0.23770371	-0.73297673	-0.031328667	0.7761412	-0.6404809	0.12914778	-0.011105422	-0.13427477	-0.06248404	-0.02266382	-0.10262683	-0.054009646	-0.16534038	0.15851502	0.09253181	0.15225035	-0.19693387	-0.21560238	-0.06921824	0.13875219	0.38441402	-0.19376493	-0.14485486	0.8504384	-0.12403153	-0.34010404	-0.57511485	0.32087132	-0.4699217	-0.5991427	-0.38354787	0.07643846	0.41233927	0.4710643	0.38365448	-0.09788056	0.290843	-0.23827966	-0.08560545	-0.13456337	0.12938945	-0.12720664	0.6335732	-0.2547157	-0.13435496	0.10237407	0.015043218	0.29236546	-0.41163802	-0.47194612	0.0021375164	-0.14972048	0.81986284	0.08114276	-0.45028347	-0.17865935	0.07186784	0.3584692	-0.16232955	-0.15823059	-0.13238655	0.044880025	0.15550476	0.47688532	0.16016406	-0.48149517	-0.071176864	-0.6077426	-0.2030059	0.1768521	-0.5485124	-0.3325962	0.3548434	0.25821534	0.9249776	0.6934977	-0.011715635	-0.14146487	0.25581717	0.20474151	-0.021625146	-0.29911506	0.13406605	-0.09719923	-0.67453307	0.60133284	0.07262376	0.23028415	0.48454234	-0.09762416	0.1699019	-0.057390846	0.04708247	-0.19954087	0.23715223	0.27060398	-0.11830914	0.14309275	0.44555667	-0.59712666	0.3388609	-0.24191463	-0.14314087	-0.7870911	-0.03023561	1.0149118	-0.3531613	-0.42143822	0.7643651	-0.12799622	-0.2331133	0.35569417	0.26482323	-0.074499995	-0.33177072	-0.36383715	0.26386356	-0.5455545	-0.008056746	-0.6303604	-0.1056522	-0.04331474	0.49403423	-0.6370182	0.46350867	-0.29608604	0.767995	0.09522082	-0.6888427	-0.25943545	-0.36920825	0.2568913	0.42724705	-0.02270616	0.38679788	0.17266603	-0.062413048	0.17743604	-0.37212738	-0.42655554	0.17206773	0.32296145	0.79271144	-0.16905421	-0.076534145	0.34987396	-0.2148987	0.9390959	0.15981968	-0.14569673	0.8177112	0.09046489	-0.29979476	0.29109263	-0.25747845	0.034589037	-0.5168783	0.29158077	0.3035596	0.27107552	-0.49266487	-0.22384566	0.59415454	0.70374495	-0.03491553	0.36812606	0.1749016	0.2222262	-0.18413563	0.7522718	-0.73904186	-0.3468923	0.35301226	0.8107471	0.6857258	0.72624457	-0.04842598	0.08171677	-0.40229577	-0.19838367	0.02009173	-0.33938697	0.5002623	-0.032842625	-0.0013138345	0.00450487	-0.7297401	-0.45924833	-0.27914205	-0.5715223	-0.038493678	-0.17521471	-0.052149385	0.0069887056	0.21819152	-0.7176853	-0.16585131	0.0024012355	-0.14961186	0.24768567	0.1144673	-0.65627795	-0.30023482	0.78784555	-0.30443558	-0.43542627	-0.95847446	-0.30929092	-0.43545228	-0.4103843	-0.18132985	0.7174908	0.4473517	0.45819798	0.30763432	-0.55492395	0.17092916	-0.18448518	0.9118716	0.33605894	0.44763833	-0.38067254	-0.13120806	0.016346814	0.61980146	-0.31204784	-0.27049953	-0.17515863	0.010215435	-0.20921604	0.31492132	-0.10566855	-0.25827888	0.60265225	-0.33362845	-0.06646537	0.89532036	0.2826628	0.104968116	-0.098076604	-0.34917298	-0.114835694	0.021354351	-0.19797084	0.08269636	-0.46286005	0.25621727	-0.044358287	0.47954494	0.57841575	-0.23782842	-0.25894716	0.21093106	0.13403569	0.38822895	0.16173702	0.0679998	0.14065406	-0.2155264	-0.8014376	0.0054304083	0.2530539	0.48012	-0.264846	0.43566376

# "new_papers_embedding.tsv" by nlp that uses default pipelines:
BMC_Med_Genomics	PMC7011249	-0.13329783	0.2985041	-0.44907343	-0.10711387	-0.27270856	-0.27364478	0.04592515	-0.32981327	0.25985417	-0.3076088	-1.0842506	-0.31346402	0.08971338	0.049379833	-0.29214936	0.21728547	0.3640917	-0.18769948	-0.22824924	0.1805722	0.18349072	-0.17979616	0.43295282	0.5300225	0.11453342	-0.20423362	-0.7133477	-0.32934496	-0.4227925	-0.12463997	0.38249305	-0.3529534	-0.43800622	-0.04097121	0.0359895	-0.23720731	-0.73272455	-0.031079803	0.7759494	-0.64112175	0.12890685	-0.010842894	-0.13427377	-0.06284978	-0.022513462	-0.102368325	-0.054214925	-0.16526884	0.1583605	0.09270608	0.1523684	-0.19713528	-0.21520148	-0.06848164	0.13854071	0.38490188	-0.19355884	-0.14503074	0.8504772	-0.12428724	-0.33967698	-0.575194	0.3211894	-0.46972448	-0.5989744	-0.38323897	0.076483205	0.41214877	0.47124505	0.38379228	-0.0981191	0.2906242	-0.2384808	-0.08514096	-0.13465413	0.12927994	-0.12736182	0.63331974	-0.2546803	-0.13434272	0.10246605	0.015231185	0.2921538	-0.41120636	-0.4712843	0.0017976207	-0.14948858	0.81935096	0.08085463	-0.45072702	-0.17916454	0.07177641	0.3586952	-0.16248855	-0.15809666	-0.13172768	0.04468074	0.15562668	0.47649956	0.16003174	-0.4813163	-0.071412496	-0.6074932	-0.202922	0.17691574	-0.5481237	-0.33243045	0.35418504	0.25869006	0.9247417	0.69340324	-0.011140373	-0.14154555	0.2558156	0.20448682	-0.02119166	-0.29957274	0.13454105	-0.09715195	-0.6750654	0.60124123	0.071792744	0.2301529	0.4845385	-0.09749041	0.17005405	-0.057398222	0.047417544	-0.19960637	0.23718034	0.27057785	-0.118096456	0.14269422	0.44519046	-0.59746414	0.33925036	-0.24228315	-0.14289501	-0.78716975	-0.030330522	1.0147663	-0.35318717	-0.42100698	0.7648233	-0.12801486	-0.23308009	0.35571018	0.26512355	-0.07438472	-0.33178163	-0.36325145	0.26371768	-0.5456856	-0.0080338335	-0.63013744	-0.105526745	-0.043356992	0.49423888	-0.6372714	0.4636167	-0.2962501	0.7680732	0.095363386	-0.68890685	-0.2592665	-0.36923698	0.25673106	0.42716873	-0.02258582	0.38716576	0.172638	-0.06250487	0.17791887	-0.3716145	-0.42696643	0.17202722	0.32335797	0.7920022	-0.16948284	-0.07673464	0.3505372	-0.21438234	0.93847114	0.15991983	-0.14584404	0.81803983	0.09041629	-0.30031037	0.2905811	-0.257404	0.033958886	-0.5169393	0.2918442	0.30326852	0.2708712	-0.49251148	-0.22351687	0.5946208	0.70332026	-0.03472063	0.36786166	0.17489213	0.22210799	-0.18444033	0.7525716	-0.7393359	-0.3471673	0.352535	0.810382	0.68553257	0.7260217	-0.04827968	0.08188952	-0.40206853	-0.19833997	0.020386191	-0.33923265	0.5002762	-0.03291202	-0.0012371549	0.004156608	-0.7295617	-0.45917565	-0.27913734	-0.5713705	-0.038378604	-0.17486218	-0.051961567	0.0074344287	0.21759564	-0.71781355	-0.1655077	0.0032356249	-0.14924853	0.24773426	0.11482885	-0.6564267	-0.30005544	0.78735924	-0.30432242	-0.43576467	-0.95888114	-0.308755	-0.43565586	-0.4103519	-0.18126269	0.71747774	0.4477098	0.45848978	0.30780977	-0.5548276	0.1705107	-0.18479058	0.9121545	0.33627686	0.44746223	-0.38045883	-0.1308282	0.016342884	0.6194594	-0.3123457	-0.27110294	-0.17504577	0.010118301	-0.20896217	0.31456918	-0.10520819	-0.25849873	0.6024002	-0.33340755	-0.06645426	0.89524645	0.28265885	0.103977345	-0.097458385	-0.34895888	-0.115495145	0.021388829	-0.19837083	0.08247762	-0.46293753	0.2564294	-0.04381285	0.47947103	0.5779067	-0.23786744	-0.2597661	0.2114751	0.13404949	0.38776726	0.16205749	0.06802507	0.1404274	-0.21495517	-0.80140054	0.0055945283	0.25279152	0.48017228	-0.26451328	0.4358428
```
3. `new_paper_token_count.tsv` is slightly different too for the same paper:
```
# "new_paper_token_count.tsv" by nlp that has no pipelines:
# ----------------------------------------------------------
PMC7011249	.	395
PMC7011249	p.	68
PMC7011249	p	21
# ...

# "new_paper_token_count.tsv" by nlp that uses default pipelines:
# -----------------------------------------------------------------
PMC7011249	.	394
PMC7011249	p.	69
PMC7011249	p	20
# ...
```
The differences must be due to the disabled pipelines. I am not an expert on NLP in spaCy. What do you think?

danich1 replied:
I'm all for disabling non-necessary features. Good find!

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,301 @@
+                    if word_counter is None:
+                        continue
+
+                    dir_writer.writerow(
```
I think this statement (lines 137-139) should be moved right after line 132, because even if `word_counter` is `None` in the current paper, the paper should be still recorded in `new_papers_added_dir.tsv`; otherwise when the updater is executed sometime in the future, this paper will be parsed again and get skipped again, which is a waste of computational resource.

dongbohu commented:
@danich1: Why the original `paper_dataset_full.tsv.xz` has 1,726,156 lines (after unzipped), but `pmc_oa_file_list.tsv` has 2,719,253 lines? I thought each line in these files should correspond to a paper that has been processed.

danich1 commented:
>@danich1: Why the original paper_dataset_full.tsv.xz has 1,726,156 lines (after unzipped), but pmc_oa_file_list.tsv has 2,719,253 lines? I thought each line in these files should correspond to a paper that has been processed.

The `paper_dataset_full.tsv.xz` only contains papers that do not have a matching preprint and are research papers. PMCOA contains documents that aren't research papers e.g. editorials or commentaries etc. Hence why papers are processed, but not included within that dataset.

cgreene commented:
Is there a reason that we exclude papers with a matching preprint?

danich1 commented:
>Is there a reason that we exclude papers with a matching preprint?

We wanted to have a subset of PMC papers that weren't apart of our golden-set when testing our KNN model. The matching preprints here are the ones within bioRxiv's xml snapshot pre-Jan 2020.

cgreene commented:
Ahh - ok! For the search server itself, it seems it would be ok to include them. Could we roll those back in so that people who have published their paper can see that the published version comes back first?

danich1 commented:
>Ahh - ok! For the search server itself, it seems it would be ok to include them. Could we roll those back in so that people who have published their paper can see that the published version comes back first?

Yeah! Sounds good!

dongbohu commented:
@danich1: How can we tell whether a paper is an editorial/commentary or a real research paper? Is all files in `comm_use.*.xml.tar.gz` the editorials/commentaries, and `non_comm_use.*.xml.tar.gz` the research papers?

danich1 commented:
>@danich1: How can we tell whether a paper is an editorial/commentary or a real research paper? Are all files in comm_use.*.xml.tar.gz the editorials/commentaries, and non_comm_use.*.xml.tar.gz the research papers?

There is an attribute within the first tag of the xml file called`article-type`. This specifies what type of article the document represents.

dongbohu commented:
@danich1 Are the papers in `comm_use.*.xml.tar.gz` and `non_comm_use.*.xml.tar.gz` supposed to be exclusive? I found that a few papers exist in both `comm_use.A-B.xml.tar.gz` and `non_comm_use.A-B.xml.tar.gz`:
- Ann_Med/PMC7544917.nxml 
- Ann_Med/PMC7784826.nxml
- Ann_Med/PMC7784832.nxml
- Ann_Med/PMC7799376.nxml
- Antonie_Van_Leeuwenhoek/PMC7415334.nxml
- Biol_Open/PMC7502593.nxml
- Biol_Open/PMC7502598.nxml
- Bone_Res/PMC7018717.nxml

I also compared the contents of `Ann_Med/PMC7544917.nxm` in both tarball files. They are not identical.

dongbohu commented:
@danich1: When `merge_files()` function merges `new_papers_embedding.tsv` into `paper_dataset_full.tsv.xz`, should we filter out editorials/commentaries from `new_papers_embedding.tsv` before merging?

danich1 commented:
>I also compared the contents of Ann_Med/PMC7544917.nxm in both tarball files. They are not identical.

Are there minor changes between those two files? I'm under the impression papers should be in either one or the other not in both. 

>@danich1: When merge_files() function merges new_papers_embedding.tsv into paper_dataset_full.tsv.xz, should we filter out editorials/commentaries from new_papers_embedding.tsv before merging?

Actually I think these types of papers should be skipped when generating the document vectors to save resources. Should be a quick check in `gather_new_papers()`

dongbohu commented:
I am attaching the two `PMC7544917.nxml` file for you to review. In order to post them in github, I compressed them into gz format:
[comm_use_PMC7544917.nxml.gz](https://github.com/greenelab/preprint-similarity-search/files/6183476/comm_use_PMC7544917.nxml.gz)
[non_comm_use_PMC7544917.nxml.gz](https://github.com/greenelab/preprint-similarity-search/files/6183477/non_comm_use_PMC7544917.nxml.gz)

I checked `gather_new_papers()` function but didn't see how editorials/commentaries are filtered out. I am not an expert on lxml. Is it done by the global variable `filter_tag_list`?

danich1 commented:
>I am attaching the two PMC7544917.nxml file for you to review. In order to post them in github, I compressed them into gz format:

These two are arguably the same based on their textual content. They don't appear to be identical because the linebreaks don't match up. One file (non_comm) has more tag line breaks than the comm version. Same authors and abstract/text and figures. Seems like a PMCOA backend mistake when adding this document.

dongbohu commented:
> These two are arguably the same based on their textual content. They don't appear to be identical because the linebreaks don't match up. One file (non_comm) has more tag line breaks than the comm version. Same authors and abstract/text and figures. Seems like a PMCOA backend mistake when adding this document.

Thanks. Just curious, what do `comm_use` and `non_comm_use` stand for?

danich1 commented:
>Thanks. Just curious, what do comm_use and non_comm_use stand for?

Commercial use (comm_use) and Non Commercial Use (non_comm). Papers have licenses on how they can be used so they are sorted into these two groups.

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,305 @@
+    all_text = root.xpath(xpath)
+    all_text = list(map(lambda x: "".join(list(x.itertext())), all_text))
+    all_text = " ".join(all_text)
+
```
@danich1 I was able to gather all new papers from the tarball files (in series, not in parallel). In order to boost the performance, I added the following code between lines 197-198:
```
    # Optimization: Remove stop words from `all_text` before feeding it to nlp.
    # This optimization not only speeds up the data processing 5%-10%, but also
    # minimizes memory usage.
    all_text = [x for x in all_text.split() if x not in nlp.Defaults.stop_words]
    all_text = " ".join(all_text)

    # Set nlp.max_length dynamically
    if nlp.max_length < len(all_text):
        nlp.max_length = len(all_text)
```
See the comments on why I made the change. Does it make sense to you?

The whole process took 5 hours and 14 minutes on my local machine. Will test it in parallel.

dongbohu replied:
@danich1: By the way, as far as I understand, it's okay to have duplicate papers in `new_papers_added_dir.tsv` ("file_path" column), but the paper IDs must be unique in `new_papers_embedding.tsv` ("document" column) and `new_paper_token_count.tsv` ("document" column), right?

danich1 replied:
>See the comments on why I made the change. Does it make sense to you?

Yeah this makes sense. I would expect a big speed up once you do things in parallel.

>@danich1: By the way, as far as I understand, it's okay to have duplicate papers in new_papers_added_dir.tsv ("file_path" column), but the paper IDs must be unique in new_papers_embedding.tsv ("document" column) and new_paper_token_count.tsv ("document" column), right?

Correct.

dongbohu commented:
A few more questions for @danich1:

1. Why the paper `PMC7047727` is only available in the original `global_doc_word_counter.tsv` but not in `paper_dataset_full.tsv.xz`?

2.  `PMC7047727` also occurred three times in the original `paper_dataset_tsne_square.tsv`:
    ```
    0.2517871260643005	-4.369272232055663	PMC7047727	2d_Mater	859
    0.2517871260643005	-4.369272232055663	PMC7047727	2d_Mater	859
    0.25178712606430054	-4.369272232055664	PMC7047727	2d_Mater	859
    ```
    Is this an error?

3. Why `global_word_counter_file` (filename: `global_doc_word_counter.tsv`) is not used after being merged? 

4. What is `global_word_counter_file_obj` (filename: `background_token_counter.pkl`)? Was it generated from "global_doc_word_counter.tsv"?

danich1 commented:
>Why the paper PMC7047727 is only available in the original global_doc_word_counter.tsv but not in paper_dataset_full.tsv.xz?

It's probably because of testing purposes as I thought I uploading a fully clean file. Turns out it probably wasn't the case. I would reset parts of the pipeline to verify the document downloader was working and forgot to reset other files. Hence the duplication error.

>PMC7047727 also occurred three times in the original paper_dataset_tsne_square.tsv:... Is this an error?

Yes. There should only be one of those IDs in the file.

>Why global_word_counter_file (filename: global_doc_word_counter.tsv) is not used after being merged?

I don't have a great way to keep track of all the tokens found within each  pmc file while simultaneously assigning each document to a specified bin. This file is needed so I don't have to process all the documents over again given that one would need to re-bin or stratify documents.

>What is global_word_counter_file_obj (filename: background_token_counter.pkl)? Was it generated from global_doc_word_counter.tsv?

Similar to above comment. Basically this object is just a big dictionary that bins documents based on the squarebin plot found on the website. This object is subjected to change if a new scheme is used or more papers get added to the model.

dongbohu commented:
Thanks @danich1. Should we remove `PMC7047727` from `global_doc_word_counter.tsv` and `paper_dataset_tsne_square.tsv`? If yes, is the removal going to affect `background_token_counter.pkl` or `bin_counters/*.pkl`?

danich1 commented:
>Should we remove PMC7047727 from global_doc_word_counter.tsv and paper_dataset_tsne_square.tsv? If yes, is the removal going to affect background_token_counter.pkl or bin_counters/*.pkl?

Yes it should be removed as it's not apart of the original dataset. Unfortunately, I'll have to say yes the pkl files will be affected as the document was in the `global_doc_word_counter.tsv`.

dongbohu commented:
@danich1: Thanks again. Can you update these files and update them to the Google cloud bucket? When you are done, I will reprocess the whole pipeline.

danich1 commented:
>@danich1: Thanks again. Can you update these files and upload them to the Google cloud bucket? When you are done, I will reprocess the whole pipeline.

Just finished uploading the data! 🤞🏾 everything runs smoothly.

dongbohu commented:
@danich1: Is `global_doc_word_counter_backup.tsv.xz` supposed to be the updated one while `global_doc_word_counter.tsv.xz` is the backup? The former was updated today but the latter was updated on March 1st.

danich1 commented:
>Is global_doc_word_counter_backup.tsv.xz supposed to be the updated one while global_doc_word_counter.tsv.xz is the backup? The former was updated today but the latter was updated on March 1st.

😤 gsutils is the worst. `global_doc_word_counter_backup.tsv.xz` is suppose to be `global_doc_word_counter.tsv.xz`. I thought I changed the name of that file. Feel free to change the back up to the correct file name.

dongbohu commented:
@danich1: I corrected the filename and put the older ones in `backup_2020-03-01` folder.

dongbohu commented:
@danich1 I checked the pickled files in `bin_counters` folder and found that they are identical to the older versions. I am not sure whether I messed up the new files that you uploaded yesterday when renaming them. Can you double check these pickle files on your local machine and tell me whether they are identical to the older ones? I used the following command to compare the two folders:
```
diff -r new_bin_counters old_bin_counters
```
If they are different on your local machine, can you create a tarball file such as "bin_counters_03_30.tar.gz" to archive the newer version and upload it to the Google cloud bucket? Thanks.

danich1 commented:
>If they are different on your local machine, can you create a tarball file such as "bin_counters_03_30.tar.gz" to archive the newer version and upload it to the Google cloud bucket? Thanks.

They are the same. Turns out I was mistaken that the word bin counters were affected. Only `paper_dataset_tsne_square.tsv` was affected as I forgot to reset that file when you pointed out the error. Everything now seems green to me.

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,305 @@
+
+        object_to_be_updated = pickle.load(
+            open(
+                f"bin_counters/word_bin_{bin_num_str + str(square_bin)}_count.pkl", "rb"
```
@danich1: After `paper_landscape_file` was updated in previous step, I found that it includes new `squarebin_id` (the last column). To give you an idea: the original `bin_counters/` directory includes pickle files whose sqaurebin_id is 0000-1477. But after `paper_landscape_file` includes the new papers, the ranges of squarebin_id's becomes 0000-1492. Because the pickle files whose squarebin_id's are 1478~1492 DO NOT exist, this line crashes the program. 

I wonder whether we should change the code into something like this:
```
        pkl_filename = f"bin_counters/word_bin_{bin_num_str + str(square_bin)}_count.pkl"
        if os.path.exists(pkl_filename):
            object_to_be_updated = pickle.load(open(pkl_filename, "rb"))
        else:
            object_to_be_updated = Counter()
```

danich1 replied:
Yes I like this implementation as it will allow for new bins to be created. Will update!

dongbohu commented on server/auto_updater/dataset_updater.py:
```diff
@@ -0,0 +1,305 @@
+            f"bin_counters/word_bin_{bin_num_str + str(square_bin)}_count.pkl"
+        )
+        if Path(pkl_filename).exists():
+            object_to_be_updated = pickle.load(open(pkl_filename, "rb"))
```
@danich1: After checking the new pickle files in `bin_counters/` folder, I feel that it may not be correct to use the original pickle files in `bin_counters/` here, because a lot of papers in original `paper_dataset_tsne_square.tsv` have been assigned to a new `squarebin_id` when `paper_dataset_tsne_square.tsv`  is updated. 
- Example 1: In the former tsne file, `PMC89010` was in squarebin **964**, but in the latter tsne file, it's in squarebin **971**.  
- Example 2: The original squarebin_id of `PMC6207739` was 1475, but was updated to 1490. Because this paper is not new, it's not never included in the updated `word_bin_1490_count.pkl`, but is still incorrectly kept in `word_bin_1475_count.pkl` after updating.

So I think we need to recalculate each squarebin's pickle file based on *ALL* papers (instead of updating the pickle files based on *new* papers only).

danich1 commented:
Updated the files to fix the reported issue. Pkl files should no longer be needed for this pipeline as the counts will be built within memory.

dongbohu commented on server/auto_updater/embedding_bin_updater.py:
```diff
@@ -0,0 +1,307 @@
+    Parameters:
+        new_papers - the file that contains all the paper embeddings
+        paper_landscape_file - the file that contains papers mapped to bins
+        paper_landscape_json_file - the json file the frontend uses
```
This parameter is not being used by this function.

dongbohu commented on server/auto_updater/embedding_bin_updater.py:
```diff
@@ -0,0 +1,307 @@
+    (
+        square_bin_plot_df.assign(bin_id=mapped_data_df.squarebin_id.unique().shape[0])[
+            ["x", "y", "xmin", "xmax", "ymin", "ymax", "bin_id"]
+        ].to_json("pmc_square_plot.json", orient="records", lines=False)
```
@danich1: Is `"pmc_square_plot.json"` on this line supposed to be `paper_landscape_json_file`?

danich1 replied:
yes. I'll change that

dongbohu commented on server/auto_updater/embedding_bin_updater.py:
```diff
@@ -0,0 +1,307 @@
+    (
+        square_plot_df.merge(pd.DataFrame.from_records(bin_stat_records), on="bin_id")
+        .reset_index(drop=True)
+        .to_json(paper_landscape_json_file, orient="records", lines=False)
```
Just want to confirm: so `paper_landscape_json_file` is created by `generate_SAUCIE_coordinates()` earlier, then modified here?

danich1 replied:
Correct

dongbohu commented on server/auto_updater/embedding_bin_updater.py:
```diff
@@ -0,0 +1,307 @@
+        if line_num == 0:
+            dim = len(line) - 2
+
+        bin_key = doc_bin_mapper[line["document"]]
```
What is `doc_bin_mapper`?

danich1 replied:
It's suppose to be a dictionary that has document ids and keys and the bins they are assigned as values. I forgot to pass that object along to the function, so I fixed it.

dongbohu commented on server/auto_updater/embedding_bin_updater.py:
```diff
@@ -0,0 +1,309 @@
+
+    # Write updated bin plot to disk
+    (
+        square_bin_plot_df.assign(bin_id=mapped_data_df.squarebin_id.unique().shape[0])[
```
@danich1: The final output json file `pmc_square_plot.json` is still empty in the latest run. I think this line has a bug:
`bin_id` should be set to the current bin ID, but `mapped_data_df.squarebin_id.unique().shape[0]` returns the total number of bins. So in the merging step (line 306 of `embedding_bin_updater.py`):
```
square_plot_df.merge(pd.DataFrame.from_records(bin_stat_records), on="bin_id")
```
No `bin_id` matches, and the resulting JSON file is empty.

(You don't have to work on this on the weekend. Please take a look at it on Monday. Thanks.)

danich1 replied:
🤦🏾 that is a silly mistake. Just remedied the problem.

dongbohu commented:
@danich1: Thank you for the new commits. I have a few more questions:
1. When `nlp` in `spacy` module parses input text, why it didn't do word stemming? For example, I expected that this line:
    ```
    nlp("came come mice mouse")
    ``` 
    would generate tokens of only `come` and `mouse`, but it actually includes all four input words.

2. In the original `global_doc_word_counter.tsv`, why many lines have identical `document` and `lemma` values. For example, the following lines:
```
PMC2584016	mice	3
PMC2584016	mice	29
PMC2584016	mice	26
PMC2584016	mice	2
PMC2584016	mice	1
PMC2584016	mice	2
PMC2584016	mice	2
```
Does it make sense to merge these lines together by adding the `count` values (3rd column)? It would decrease the size of this file and make data processing a little faster in the future.

3. In the original `paper_dataset_full.tsv`, some papers have identical `feat_n` values (columns 3 to 302). In other words, the distances between any pairs of these papers are 0. For example, the following papers have identical feat values:
```
Br_J_Cancer	PMC1968158
Br_J_Cancer	PMC1968162
Br_J_Cancer	PMC1968163
Br_J_Cancer	PMC1968175
Br_J_Cancer	PMC1968189
Br_J_Cancer	PMC1968196
Br_J_Cancer	PMC1968197
Br_J_Cancer	PMC1968198
Br_J_Cancer	PMC1968199
Br_J_Cancer	PMC1968203
```
Is it by accident?

dongbohu commented:
One more question for @danich1: Why the original `global_doc_word_counter.tsv` includes many lines whose lemma column (column 2) is blank? For example:
```
PMC4350859	"				"	18
```
Should we delete these lines?

danich1 commented:
>would generate tokens of only come and mouse, but it actually includes all four input words.

The lemmatizer is disabled for speed so as a result the words aren't in their stemmed form. Will need to enable the lemmatizer for words to be in there stem form. 

>In the original global_doc_word_counter.tsv, why many lines have identical document and lemma values?

My initial hunch here is that these words contain some non-alphanumeric unicode character(s), which result in the mice tokens being unique when they are indeed the same.

>Is it by accident?

Great question. It's either the papers are using the exact same word vectors for their embedding calculations or this is an accident. I'm leaning towards the latter as using the exact same word vectors for all those papers seems highly unlikely. 

>Why the original global_doc_word_counter.tsv includes many lines whose lemma column (column 2) is blank?

Basically, those spaces are actually a series of tab characters that should be removed.

danich1 commented:
Overall, I need to re-run everything as the parser I was using for the original plot is quite outdated. Plus, I need to add more papers to the paper_dataset and a few other todos. I'll update you when re-running has finished. Other than that we are quite close!!

dongbohu commented:
@danich1 Thank for your reply.
> The lemmatizer is disabled for speed so as a result the words aren't in their stemmed form. Will need to enable the lemmatizer for words to be in there stem form.

I tried the following code:
```
import spacy
nlp =  spacy.load("en_core_web_sm")
print(nlp.pipe_names)
``` 
The output is:
```
['tok2vec', 'tagger', 'parser', 'ner', 'attribute_ruler', 'lemmatizer']
```
You can see that `lemmatizer` is enabled in the pipelines. But the tokens generated by `nlp("come came")` still include both `come` and `came`. Why is that?

danich1 commented:
>You can see that lemmatizer is enabled in the pipelines. But the tokens generated by nlp("come came") still include both come and came. Why is that?

Spacy returns the same parsed string as input but if you were to do:
```python
parsed_sentence = nlp("come came")
for tok in parsed_sentence:
   print(tok.lemma_)
```

Then your output would be "come come".

dongbohu commented:
@danich1 Good to know. Does that mean we should use `.lemma_` in this function?
https://github.com/danich1/preprint-similarity-search/blob/f9c1aa1ecc86cb39ff152f725d2f7c87aa1856ec/server/auto_updater/dataset_updater.py#L210

danich1 commented:
>Good to know. Does that mean we should use .lemma_ in this function?

Yes. I'm including in the re-run. Once that is done I'll updated everything accordingly.

dongbohu commented:
@danich1 In case you didn't know: I did a few tests. The default `nlp` pipelines in `spacy` are:
- "tok2vec"
- "tagger"
- "parser"
- "ner"
- "attribute_ruler"
- "lemmatizer"

"parser" and "ner" can be disabled without affecting `lemma_`. You can confirm it.

danich1 commented:
>By the way, when you generated the updated paper_dataset_full.tsv.xz file, did you disable any pipelines?

Yeah I did. "Tok2vec", "attribute_ruler" and "Tagger" can be disabled as well. (ran with those disabled and no warnings or issues were fired.

dongbohu commented:
@danich1 When running the following code:
```
disabled_pipelines = [
    "tok2vec", 
    "tagger",
    "parser",
    "ner",
    "attribute_ruler",
]
nlp = spacy.load("en_core_web_sm", disable=disabled_pipelines)
foo = nlp("went came mice")
```
I got the following error:
```
[W108] The rule-based lemmatizer did not find POS annotation for the token 'went'. Check that your pipeline includes components that assign token.pos, typically 'tagger'+'attribute_ruler' or 'morphologizer'.
[W108] The rule-based lemmatizer did not find POS annotation for the token 'came'. Check that your pipeline includes components that assign token.pos, typically 'tagger'+'attribute_ruler' or 'morphologizer'.
[W108] The rule-based lemmatizer did not find POS annotation for the token 'mice'. Check that your pipeline includes components that assign token.pos, typically 'tagger'+'attribute_ruler' or 'morphologizer'.
```

danich1 commented:
Right... Tagger is needed as the POS tags are necessary to lemmatize words. `paper_dataset_full.tsv.xz` was generated with tagger on but the other pipelines I mentioned removed. Just confirmed with the code I currently have.

dongbohu commented:
Hmm, I changed my example to:
```
disabled_pipelines = ['tok2vec', 'parser', 'ner', 'attribute_ruler']
```
but still got the same errors. (By the way, I am using Python 3.8.5.)

danich1 commented:
>but still got the same errors. (By the way, I am using Python 3.8.5.)

Yeah never mind you are right. Only parser and ner can be disabled. Ignore my messages above.

dongbohu commented:
Thanks. Please don't forget to update `disabled_pipelines` variable.

dongbohu commented:
@danich1 I chcked the files that you uploaded yesterday and have a few questions:
1. Why `paper_dataset_full.tsv` includes 1,851,724 papers, but `paper_dataset_tsne_square.tsv` has `1,742,067` papers? Shouldn't the papers be identical in these two files?

2. In `global_doc_word_counter.tsv`, many words don't seem to be converted to their stems. For example, I see these lines:
```
PMC2714472	currently	3
PMC2714472	contains	2
PMC2714472	representing	2
PMC2714472	plants	1
PMC2714472	animals	1
```
Why these words are not converted to the stems such as `current`, `contain`, `represent`, `plant`, `animal`?

dongbohu commented:
One more question: 

Originally `paper_dataset_full.tsv` includes 302 columns (`journal`, `document`, and 300 `feat_n` columns). But the new one includes 301 columns now, and `journal` column is gone. Is there any reason why? 

This will make the calculation of centroid a little harder, because we have to use another file (such as `paper_dataset_tsne_square.tsv`) to build a map between a paper and its corresponding journal.

danich1 commented:
>Originally paper_dataset_full.tsv includes 302 columns (journal, document, and 300 feat_n columns). But the new one includes 301 columns now, and journal column is gone. Is there any reason why?

Wrong file was copied, which is what caused this discrepancy. I'm uploading the correct one now.

danich1 commented:
>Why paper_dataset_full.tsv includes 1,851,724 papers, but paper_dataset_tsne_square.tsv has 1,742,067 papers? Shouldn't the papers be identical in these two files?

It goes back to that wrong file issue. When I was trying to copy files over to run a certain notebook, I was using an older version of the file. Hence the discrepancy as mentioned previously.

>Why these words are not converted to the stems such as current, contain, represent, plant, animal?

The tokens aren't going to be changed for this part as the lemmatizer is only used for generating the document vectors itself. Counter argument against lemmatizer in this part is that words like "fig" and "figure" wouldn't change once used. Were you expected me to use the lemmatizer on the tokens themselves or just for document vectors?

dongbohu commented:
> The tokens aren't going to be changed for this part as the lemmatizer is only used for generating the document vectors itself. Counter argument against lemmatizer in this part is that words like "fig" and "figure" wouldn't change once used. Were you expected me to use the lemmatizer on the tokens themselves or just for document vectors?

Yes, I did expect that the token counters would use lemmatizer as well. Since the distance is calculated based on the lemmatizer, wouldn't that make more sense?

On the other hand, if we do use lemmatizer only for word vector calculation, then I think this part: https://github.com/greenelab/preprint-similarity-search/blob/b9a199c4b7f70e34681a62569a3ed7840676c6a4/server/auto_updater/dataset_updater.py#L237
should be changed, because the token counter is using lemmatized `all_tokens`.

dongbohu commented:
By the way, another advantage of using lemmatizer in token counter is that the file size of `global_doc_word_counter.tsv` will be much smaller, and the data processing will be faster (I don't know how much though).

danich1 commented:
>Yes, I did expect that the token counters would use lemmatizer as well. Since the distance is calculated based on the lemmatizer, wouldn't that make more sense?

The lemmatizer wouldn't make a significant change in the global picture as quick glance within the token bins contains mostly unique tokens; however, to be through running a lemmatizer would make the most sense.  Means it's a rerun again...

dongbohu commented:
@danich1 I am sure that you know much more than I do on this topic. So I leave the decision to you and @cgreene.

cgreene commented:
I think that for the first part of the work (Figure 1), we really should be using the tokens because we may see changes in the language itself through refinement (e.g., Figure / Fig as David mentions) and that analysis is trying to describe the language as used. For the work beyond that, I think we benefit from having the semantic elements intact (as opposed to the precise tokens). I'm not sure how that relates to the above questions though, since they refer to the server.

@danich1: if you are doing a bunch of copying files around with multiple versions we should figure out if git-lfs or something else should be used for recordkeeping. There needs to be something standard to avoid mismatched versions and to know which version is used for analyses.

dongbohu commented:
@danich1 If we use the tokens w/o lemmatizer when calculating the counters, I think https://github.com/greenelab/preprint-similarity-search/blob/b9a199c4b7f70e34681a62569a3ed7840676c6a4/server/auto_updater/dataset_updater.py#L237-L254 would be written like this:
```
    valid_tokens = filter(
        lambda tok: str(tok.lemma_) in model.wv and str(tok.lemma_) not in nlp.Defaults.stop_words,
        nlp(all_text)
    )

    lemmas = list(map(lambda x: x.lemma_, valid_tokens))
    word_vectors += [model.wv[w] for w in lemmas]

    # skips weird documents that don't contain text
    if len(word_vectors) > 0:
        return np.stack(word_vectors).mean(axis=0), Counter(valid_tokens)

    return [], None
```
Does this make sense to you?

danich1 commented:
>Does this make sense to you? 

Yeah that makes sense.

danich1 commented:
Just a heads up I decided I'll run the lemmatizer for this part of the project and for document vector generation. It will improve the token association (albeit minor) for each bin on the publication map.  Should be done by Monday 🤞🏾 along with the correct files uploaded for testing the auto updater.

dongbohu commented:
@danich1 If you rerun the whole process with lemmatizer for both word vector calculation and token counters, that means the auto-updater should use lemmatizer for both too, right?

danich1 commented:
> @danich1 If you rerun the whole process with lemmatizer for both word vector calculation and token counters, that means the auto-updater should use lemmatizer for both too, right?

Yup. That is correct!

danich1 commented:
Alright! Everything should be in order. I have uploaded the new data onto google cloud updated-data folder and I updated the code to fix the spaCy parser (lemmatizer wasn't working right as too many pipelines were disabled). I also updated the necessary code/plot files for #89.

dongbohu commented:
@danich1: I checked the latest files that you uploaded on Friday and have a few questions about the file `global_doc_word_counter.tsv`:

1. Why some papers at the beginning of this file only include very few tokens (one or two):
   ```
   document	lemma	count
   PMC2557413	image	1
   PMC2557413	   	1
   PMC2557483	image	1
   PMC2530993	image	1
   PMC2530993	   	1
   PMC2530995	image	1
   PMC2530995	  	1
   PMC2530991	image	1
   PMC2530991	 	1
   PMC2642002	image	1
   PMC2642002	 	1
   PMC2530973	image	1
   PMC2530973	 	1
   PMC2557420	image	1
   PMC2557420	  	1
   PMC2531019	image	1
   ... ...
  ```
   Do you think these papers should be excluded?

2. Some tokens don't seem to be lemmatized correctly, for example:
   ```
   PMC1533399	officially	1
   PMC1533399	generally	1
   PMC1533399	thinking	1
   ```
   Any idea why? Thanks.

danich1 commented:
>Why some papers at the beginning of this file only include very few tokens (one or two): ... Do you think these papers should be excluded?

These papers are literal scans of papers back in the early 1900s, before digital copies existed. I say keep them for record keeping and consistency. They don't really affect any of the results or anything.

>Some tokens don't seem to be lemmatized correctly, for example: 

That is a spaCy issue with their model. Words such as you described above don't get lemmatized because of their backend model. I doubled checked on my own and I get the same thing with the lemmatizer enabled/disabled. No reasonable fix unfortunately.

dongbohu commented:
@danich1 Thanks for your reply. I still don't understand why "thinking" was not converted to "think" correctly in `global_doc_word_counter.tsv`. But I am starting the auto-updater pipeline anyway.

dongbohu commented on server/auto_updater/embedding_bin_updater.py:
```diff
@@ -43,6 +43,7 @@ def generate_centroid_dataset(paper_dataset_file, output_file="centroid_dataset.
         if line["journal"] not in journal_centroid:
             journal_centroid[line["journal"]] = {
                 "journal": line["journal"],
+                "document": line["document"],
```
@danich1 So this means the representative paper in a journal will be always the first paper in `dataset_dataset_full.tsv`?

danich1 replied:
Sorry was in a car all day. 
Yeah. That's what i originally had as a representative paper.

danich1 commented:
>RuntimeError: dictionary changed size during iteration

What line gave you that error or did you fix this issue?

dongbohu commented:
> What line gave you that error or did you fix this issue?
The error was due to `journal_centroid["document"]`, so it's fixed by your last commit.
